src/sv_snapshot.c

```diff
@@ -106,233 +106,173 @@ void SV_UpdateServerCommandsToClientRecover( client_t *client, msg_t *msg )
-        oldframe = NULL;
-        lastframe = 0;
-        var_x = 0;
-
-    } else if(client->demoDeltaFrameCount <= 0 && client->demorecording){
-
-        oldframe = NULL;
-        lastframe = 0;
-        var_x = 0;
-        client->demowaiting = qfalse;
-        Com_DPrintf(CON_CHANNEL_SERVER,"Force a nondelta frame for %s for demo recording\n", client->name);
-
-        if(client->demoMaxDeltaFrames < 1024)
-        {
-            client->demoMaxDeltaFrames <<= 1;
-        }
-        client->demoDeltaFrameCount = client->demoMaxDeltaFrames;
-
-
-    } else {
-        oldframe = &client->frames[client->deltaMessage & PACKET_MASK];
-        lastframe = client->netchan.outgoingSequence - client->deltaMessage;
-        var_x = oldframe->var_03;
-        client->demoDeltaFrameCount--;
-
-        if(oldframe->first_entity <  svsHeader.nextSnapshotEntities - svsHeader.numSnapshotEntities) {
-            Com_DPrintWarning(CON_CHANNEL_SERVER,"%s: Delta request from out of date entities - delta against entity %i, oldest is %i, current is %i.  Their old snapshot had %i entities in it\n",
-                            client->name, oldframe->first_entity, svs.nextSnapshotEntities - svs.numSnapshotEntities, svs.nextSnapshotEntities, oldframe->num_entities );
-            oldframe = NULL;
-            lastframe = 0;
-            var_x = 0;
-
-        } else if(oldframe->first_client <  svsHeader.nextSnapshotClients - svsHeader.numSnapshotClients) {
-
-            Com_DPrintWarning(CON_CHANNEL_SERVER,"%s: Delta request from out of date clients - delta against client %i, oldest is %i, current is %i.  Their old snapshot had %i clients in it\n",
-                            client->name, oldframe->first_client, svs.nextSnapshotClients - svs.numSnapshotClients, svs.nextSnapshotClients, oldframe->num_clients);
-            oldframe = NULL;
-            lastframe = 0;
-            var_x = 0;
-        }
-    }
-
-
-    MSG_WriteByte(msg, svc_snapshot);
-    MSG_WriteLong(msg, svsHeader.time);
-    MSG_WriteByte(msg, lastframe);
-    snapInfo.snapshotDeltaTime = var_x;
-
-    snapFlags = svsHeader.snapFlagServerBit;
-
-    if(client->rateDelayed){
-	snapFlags |= 1;
-    }
+__cdecl void SV_WriteSnapshotToClient(client_t* client, msg_t* msg)
+{
+	qboolean nonDeltaFrameRequested = qfalse;
+	qboolean deltaFrame = qfalse;
 
-    if(client->state == CS_ACTIVE) {
+	if (client->demorecording && client->demowaiting)
+	{
+		// store the sequence when the demo was requested to start
+		if (client->demoRequestedSequence <= 0)
+		{
+			client->demoRequestedSequence = client->netchan.outgoingSequence;
+		}
 
-	client->unksnapshotvar = 1;
+		// if the client is up-to-date, or the demo has been waiting for PACKET_BACKUP (32) or more snapshots, start the demo with a non-delta snapshot
+		if (client->netchan.outgoingSequence - client->deltaMessage == 1 || client->netchan.outgoingSequence - client->demoRequestedSequence >= PACKET_BACKUP)
```

Comment:
`client->netchan.outgoingSequence - client->deltaMessage == 1`
I've no idea if this works correctly, but the idea is to find some logic that limits the maximum number of non-delta snapshots / frames to the absolute minimum, just one.